### PR TITLE
[SYCL] Add BFloat16 aspect

### PR DESCRIPTION
**Pull request by MrSidims**
Currently this aspect is mapped on a device info which is
being proved by get_platform_name function, claiming, that any
Intel GPU device supports BFloat16 extension. Once we have
a lower-level spec we can re-map it on the appropriate OpenCL
extension.

Spec:
https://github.com/intel/llvm/pull/4237

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

File: sycl/source/detail/device_info.hpp [hunk 1]
```diff
@@ -493,6 +493,25 @@ template <> struct get_device_info<device, info::device::parent_device> {
   }
 };
 
+// Specialization for bfloat16
+template <>
+struct get_device_info<bool, info::device::ext_intel_bf16_conversion> {
+  static bool get(RT::PiDevice dev, const plugin &Plugin) {
+    // TODO: We claim, that all Intel GPU devices support bfloat16 feature
```
Comment (AlexeySotkin) [inline sycl/source/detail/device_info.hpp:500]: "support bfloat16 feature" -> "support bfloat16 conversion feature"

Comment (MrSidims) [reply sycl/source/detail/device_info.hpp]: Done

File: sycl/source/detail/device_info.hpp [hunk 2]
```diff
@@ -493,6 +493,23 @@ template <> struct get_device_info<device, info::device::parent_device> {
   }
 };
 
+// Specialization for bfloat16
+template <>
+struct get_device_info<bool, info::device::ext_intel_bf16_conversion> {
+  static bool get(RT::PiDevice dev, const plugin &Plugin) {
+    // TODO: We claim, that all Intel GPU devices support bfloat16 conversion
+    // feature but we'd better have a low-level extension to query for support
+    // of the feature
```
Comment (s-kanaev) [inline sycl/source/detail/device_info.hpp:502]: Is there any spec for the extension?

File: sycl/source/detail/device_info.hpp [hunk 3]
```diff
@@ -493,6 +493,23 @@ template <> struct get_device_info<device, info::device::parent_device> {
...
+    platform plt =
+        get_device_info<platform, info::device::platform>::get(dev, Plugin);
+    std::string platform_name = plt.get_info<info::platform::name>();
+    if (platform_name == "Intel(R) OpenCL HD Graphics")
```
Comment (s-kanaev) [inline sycl/source/detail/device_info.hpp:506]: I believe, there should be a better way to verify the device is either of Intel's GPUs.

**Review by s-kanaev — COMMENTED** [review]
Although the changes look good to me, @romanovvlad  do we allow for new features in Runtime nowadays?

File: sycl/source/detail/device_info.hpp (hunk 2 above)
Comment (MrSidims) [reply sycl/source/detail/device_info.hpp:502]: Here it is https://github.com/intel/llvm/pull/4237

File: sycl/source/detail/device_info.hpp (hunk 3 above)
Comment (MrSidims) [reply sycl/source/detail/device_info.hpp:506]: Right now there is no low-level runtime specification for the feature, so it's the best, what came in my mind. Do you have any ideas?

File: sycl/include/CL/sycl/info/info_desc.hpp [hunk 4]
```diff
@@ -154,6 +154,7 @@ enum class device : cl_device_info {
   atomic64 = PI_DEVICE_INFO_ATOMIC_64,
   atomic_memory_order_capabilities =
       PI_DEVICE_INFO_ATOMIC_MEMORY_ORDER_CAPABILITIES,
+  ext_intel_bf16_conversion,
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/info/info_desc.hpp:157]: Are we going to map it to a specific value in future?

File: sycl/source/detail/device_info.hpp (hunk 2 above)
Comment (romanovvlad) [reply sycl/source/detail/device_info.hpp:502]: It seems #4237 is a spec for SYCL extension. Is there any spec for "a low-level extension"?

Could you please clarify why the temporary solution thru checking platform name is needed? Can we wait and implement the proper solution?

**Review by romanovvlad — COMMENTED** [review]
> @romanovvlad do we allow for new features in Runtime nowadays?

Yes, we do.

File: sycl/include/CL/sycl/info/info_desc.hpp (hunk 4 above)
Comment (MrSidims) [reply sycl/include/CL/sycl/info/info_desc.hpp:157]: Yes

**Comment by MrSidims** [comment]
>> It seems #4237 is a spec for SYCL extension. Is there any spec for "a low-level extension"?

Nothing to share right now

>> Could you please clarify why the temporary solution thru checking platform name is needed? Can we wait and implement the proper solution?

So users could already write bfloat16 - sensitive code using aspect traits deciding for them which version of kernel should be enqueued: with or without bf16 conversion feature.

File: sycl/source/detail/device_info.hpp (hunk 2 above)
Comment (MrSidims) [reply sycl/source/detail/device_info.hpp:502]: Coping my answer from below.

>> It seems #4237 is a spec for SYCL extension. Is there any spec for "a low-level extension"?

Nothing to share right now

>> Could you please clarify why the temporary solution thru checking platform name is needed? Can we wait and implement the proper solution?

So users could already write bfloat16 - sensitive code using aspect traits deciding for them which version of kernel should be enqueued: with or without bf16 conversion feature.

**Comment by MrSidims** [comment]
E2E test: https://github.com/intel/llvm-test-suite/pull/390

**Comment by MrSidims** [comment]
Closing: there are several issues with not mapping an aspect directly on a low-level extension. Also to use aspects for this device code feature we also need https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DeviceIf/device_if.asciidoc which is not implemented.

**Pull request closed without merging**